### PR TITLE
#187300257 buyer is able to create a cart.

### PR DESCRIPTION
**Pull request by ErnestTchami**
#### What does this PR do?

This PR enhances the functionality related to user shopping carts. It enables buyers to create new carts and update existing ones.

#### Endpoints
-POST : /api/carts   
-GET:  /api/carts 
-PUT : /api/carts 

#### Description of Task to be completed?
Implement functionality to handle cart operations such as creation, retrieval, and updating for buyers.

#### How Should This be Manually Tested?
- By use postman 
- swagger

#### What are the relevant pivotal tracker/Trello stories?
- [pivotal id #187300257](https://www.pivotaltracker.com/n/projects/2698211)
#### Screenshots
image

### CODE REVIEW

**Review by emmanueltct — APPROVED** [review]

**Review by kayigmb — APPROVED** [review]

File: src/__test__/cart.spec.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
+// import { describe, it } from "mocha";
+// import chai, { expect } from "chai";
```
Comment (Eli250) [inline src/__test__/cart.spec.ts]: Why is the entire file commented?

File: src/controllers/cart.controller.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  getProductsWithQuantity
+} from "../utils/cartMergeProduct";
+
+interface CartItem {
+  productId: string;
+  Quantity: number;
+}
+export const createCart = async (req: Request, res: Response) => {
+  try {
+    const newuser: any = req.user;
```
Comment (Eli250) [inline src/controllers/cart.controller.ts]: Do not use any

File: src/controllers/cart.controller.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    let sum = 0;
+    productsWithQuantity.map((el) => {
+      sum += el.totalPrice;
+    });
+
+    const news = await mergeDuplicatedProduct(productsWithQuantity);
+    if (!news.StockCheck) {
+      return res.status(404).json({
+        error:
+          "you send the quantity which are more than what we have in the stock Please check."
```
Comment (Eli250) [inline src/controllers/cart.controller.ts]: What does this message mean? What are you asking the user to check?

File: src/controllers/cart.controller.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    const newuser: any = req.user;
+    const cartItems: CartItem[] = req.body;
+
+    const productsWithQuantity = await getProductsWithQuantity(cartItems);
+    let sum = 0;
+    productsWithQuantity.map((el) => {
+      sum += el.totalPrice;
+    });
+
+    const news = await mergeDuplicatedProduct(productsWithQuantity);
```
Comment (Eli250) [inline src/controllers/cart.controller.ts:25]: This variable name is ambiguous.

File: src/controllers/cart.controller.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      status: error.status,
+      message: "the request fail",
+      error: error.message
+    });
+  }
+};
+
+export const getCart = async (req: Request, res: Response) => {
+  try {
+    const newuser: any = req.user;
```
Comment (Eli250) [inline src/controllers/cart.controller.ts]: Don't use any

File: src/controllers/cart.controller.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  } catch (error: Error | any) {
+    res
+      .status(500)
+      .json({ message: "internal server error", error: error.message });
+  }
+};
+
+export const updateCart = async (req: Request, res: Response) => {
+  try {
+    const newuser: any = req.user;
```
Comment (Eli250) [inline src/controllers/cart.controller.ts]: Don't use any

File: src/middlewares/checkusercart.middleware.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+export const isUserWhoNothaveCart = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const newuser: any = req.user;
+  const new_cart = await Cart.findOne({ where: { userId: newuser.id } });
+  if (!new_cart) {
+    return res.status(404).json({
+      error: "Please create cart by post the product at the first time"
```
Comment (Eli250) [inline src/middlewares/checkusercart.middleware.ts:28]: This message is unclear.

**Review by Eli250 — CHANGES_REQUESTED** [review]

File: src/__test__/cart.spec.ts (hunk 1 above)
Comment (ErnestTchami) [reply src/__test__/cart.spec.ts]: I needed a separate file to test my endpoint, but it required me to duplicate some code from the product. Instead, I chose to use the same file with the product. But I forgot to delete the cart file. The issue is now resolved.

File: src/controllers/cart.controller.ts (hunk 3 above)
Comment (ErnestTchami) [reply src/controllers/cart.controller.ts]: The issue is now resolved.

File: src/controllers/cart.controller.ts (hunk 4 above)
Comment (ErnestTchami) [reply src/controllers/cart.controller.ts:25]: The issue is now resolved.

File: src/controllers/cart.controller.ts (hunk 5 above)
Comment (ErnestTchami) [reply src/controllers/cart.controller.ts]: The issue is now resolved

File: src/controllers/cart.controller.ts (hunk 6 above)
Comment (ErnestTchami) [reply src/controllers/cart.controller.ts]: The issue is now resolved

**Review by istafrica-dennw — APPROVED** [review]

**Pull request merged**